lib/hammer_cli/defaults.rb

```diff
@@ -0,0 +1,143 @@
+require 'hammer_cli'
+require 'yaml'
+module HammerCLI
+
+  class DefaultsCommand < HammerCLI::AbstractCommand
+
+    class ListDefaultsCommand < HammerCLI::DefaultsCommand
+      command_name 'list'
+
+      desc _('List all the defaults parameters')
+      def execute
+        list_all_defaults_message
+        HammerCLI::EX_OK
+      end
+    end
+
+    class AddDefaultsCommand < HammerCLI::DefaultsCommand
+      command_name 'add'
+
+      desc _('Add a default parameter to config')
+      option "--param-name", "OPTION_NAME", _("The name of the default option"), :required => true
+      option "--param-val", "OPTION_VALUE", _("The value for the default option")
+      option "--plugin-name", "OPTION_PLUGIN_NAME", _("The plugin name defaults will be generated for")
+
+      def execute
+        begin
+          if option_plugin_name.nil? && option_param_val.nil? || !option_plugin_name.nil? && !option_param_val.nil?
+            bad_input
+          else
+            if option_plugin_name
+              namespace = ("HammerCLI" + option_plugin_name.split('_').collect!{ |w| w.downcase.capitalize }.join + "::Defaults").to_sym
+              raise NameError unless HammerCLI::Defaults.providers.keys.include?(namespace)
```

Comment:
Looking at the namespace, one improvement came to my mind:
What about registering the providers exactly with the same name that is expected in the parameter? I think that would simplify the namespace manipulation. Example:

```ruby
class BaseDefaultsProvider
  def name
    self.class.name.split('::').first.gsub(/^HammerCLI/, '').underscore
    # ...or some other reasonable default
  end
  # not implmented exception for
  # def support?
  # def get_defaults
end

module HammerCLIForeman
  class DefaultsProvider < BaseDefaultsProvider
    # def support?
    # def get_defaults
    # ...
  end
end

HammerCLI::Defaults.register_provider(DefaultsProvider.new)
```

Than you can simply use `HammerCLI::Defaults.providers.key?(provider_name)`. It will also make the `Defaults` class more testable plus we won't restrict the names of the provider classes at all.